solver/src/encoding.rs

```diff
@@ -41,23 +43,37 @@ fn encode_order_flags(order: &OrderCreation) -> u8 {
     result
 }
 
-fn encode_uniswap_call(
-    amount_in: U256,
-    amount_out_min: U256,
-    token_in: H160,
-    token_out: H160,
-    payout_to: H160,
-) -> [u8; UNISWAP_INTERACTION_SIZE] {
+pub fn encode_uniswap_call(
+    target: &H160,
+    amount_in: &U256,
+    amount_out_min: &U256,
+    token_in: &H160,
+    token_out: &H160,
+    payout_to: &H160,
+) -> [u8; UNISWAP_TOTAL_SIZE] {
     let uniswap = UniswapV2Router02::at(&dummy::dummy_web3(), H160::zero());
     let method = uniswap.swap_exact_tokens_for_tokens(
-        amount_in,
-        amount_out_min,
-        vec![token_in, token_out],
-        payout_to,
+        *amount_in,
+        *amount_out_min,
+        vec![*token_in, *token_out],
+        *payout_to,
         U256::MAX,
     );
     let data = method.tx.data.expect("call doesn't have calldata").0;
-    data.try_into().expect("unexpected data size")
+    let mut result = [0u8; UNISWAP_TOTAL_SIZE];
+    result[..20].copy_from_slice(target.as_fixed_bytes());
+    result[20..23].copy_from_slice(&encode_interaction_data_length(UNISWAP_DATA_SIZE).unwrap());
+    result[23..].copy_from_slice(data.as_slice());
+    result
+}
+
+fn encode_interaction_data_length(length: usize) -> Option<[u8; 3]> {
+    let bytes = length.to_be_bytes();
+    let (left, right) = bytes.split_at(bytes.len() - 3);
+    if left.iter().any(|byte| *byte != 0) {
+        return None;
+    }
+    Some(right.try_into().expect("usize smaller than 3 bytes"))
```

Comment:
I didn't fully understand the expect message at first. It cannot fair because we split_at length - 3, right? Maybe saying 'right has exactly 3 bytes' ?

ubernit: Maybe just if length > 2**12 then instead of iterating (or adding a comment that this checks we don't overflow)

Reply:
Added a comment to explain this more.